File: app/src/main/java/top/easelink/lcg/ui/main/articles/viewmodel/FavoriteArticlesAdapter.java
package top.easelink.lcg.ui.main.articles.viewmodel;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;
import top.easelink.framework.base.BaseViewHolder;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.ItemFavoriteArticleViewBinding;
import top.easelink.lcg.ui.main.source.local.ArticlesLocalDataSource;
import top.easelink.lcg.ui.main.source.model.ArticleEntity;

import static top.easelink.lcg.ui.main.articles.viewmodel.ArticlesViewModel.FETCH_MORE;

public class FavoriteArticlesAdapter extends RecyclerView.Adapter<BaseViewHolder> {

    private static final int VIEW_TYPE_EMPTY = 0;
    private static final int VIEW_TYPE_NORMAL = 1;
    private static final int VIEW_TYPE_LOAD_MORE = 2;

    private List<ArticleEntity> mArticleEntities = new ArrayList<>();

    private ArticlesAdapterListener mListener;

    public FavoriteArticlesAdapter(ArticlesAdapterListener listener) {
        mListener = listener;
    }

    @BindingAdapter({"adapter"})
    public static void addFavoriteArticleItems(RecyclerView recyclerView, List<ArticleEntity> articleEntities) {
        FavoriteArticlesAdapter adapter = (FavoriteArticlesAdapter) recyclerView.getAdapter();
        if (adapter != null) {
            adapter.clearItems();
            adapter.addItems(articleEntities);
        }
    }

    @Override
    public int getItemCount() {
        if (mArticleEntities.isEmpty()) {
            return 1;
        } else {
            return mArticleEntities.size() + 1;
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (mArticleEntities.isEmpty()) {
            return VIEW_TYPE_EMPTY;
        } else {
            if (position == mArticleEntities.size()) {
                return VIEW_TYPE_LOAD_MORE;
            }
            return VIEW_TYPE_NORMAL;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
        holder.onBind(position);
    }

    @Override
    @NonNull
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case VIEW_TYPE_NORMAL:
                ItemFavoriteArticleViewBinding favoriteArticleViewBinding
                        = ItemFavoriteArticleViewBinding.inflate(LayoutInflater.from(parent.getContext()),
                        parent, false);
                return new ArticleViewHolder(favoriteArticleViewBinding, this);
            case VIEW_TYPE_LOAD_MORE:
                return new LoadMoreViewHolder(
                        LayoutInflater.from(parent.getContext())
                                .inflate(R.layout.item_load_more_view
                                        , parent, false));
            case VIEW_TYPE_EMPTY:
            default:
                return new EmptyViewHolder(
                        LayoutInflater.from(parent.getContext())
                                .inflate(R.layout.item_favorite_article_empty_view,
                                        parent, false));
        }
    }

    private void addItems(List<ArticleEntity> articleEntityList) {
        mArticleEntities.addAll(articleEntityList);
        notifyDataSetChanged();
    }

    private void clearItems() {
        mArticleEntities.clear();
    }

    public void setListener(ArticlesAdapterListener listener) {
        this.mListener = listener;
    }

    public interface ArticlesAdapterListener {
        int FETCH_INIT = 0;
        int FETCH_MORE = 1;
        void fetchArticles(int type);
    }

    public class ArticleViewHolder extends BaseViewHolder {

        private ItemFavoriteArticleViewBinding mBinding;
        private FavoriteArticlesAdapter mAdapter;

        private FavoriteArticleItemViewModel favoriteArticleItemViewModel;

        ArticleViewHolder(ItemFavoriteArticleViewBinding binding, FavoriteArticlesAdapter adapter) {
            super(binding.getRoot());
            this.mBinding = binding;
            this.mAdapter = adapter;
        }

        @SuppressLint("CheckResult")
        @Override
        public void onBind(int position) {
            final ArticleEntity articleEntity = mArticleEntities.get(position);
            mBinding.removeButton.setOnClickListener(v -> ArticlesLocalDataSource.getInstance().delArticleFromFavorite(articleEntity.getId())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(res -> {
                        if (res) {
                            mArticleEntities.remove(articleEntity);
                            notifyItemRemoved(position);
                            notifyItemRangeChanged(position, mArticleEntities.size() - position);
                        }
                    })
            );
            favoriteArticleItemViewModel = new FavoriteArticleItemViewModel(articleEntity);
            try {
                mBinding.contentTextView.setHtml(articleEntity.getContent().trim());
            } catch (Exception e) {
                Timber.e(e);
            }
            mBinding.setViewModel(favoriteArticleItemViewModel);
            mBinding.executePendingBindings();
        }
    }

    public class EmptyViewHolder extends BaseViewHolder {

        EmptyViewHolder(View view) {
            super(view);
        }

        @Override
        public void onBind(int position) {
        }
    }

    public class LoadMoreViewHolder extends BaseViewHolder{

        LoadMoreViewHolder(View view) {
            super(view);
        }

        @Override
        public void onBind(int position) {
            mListener.fetchArticles(FETCH_MORE);
        }
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/articles/viewmodel/FavoriteArticlesAdapter.kt
package top.easelink.lcg.ui.main.articles.viewmodel

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.item_favorite_article_empty_view.view.*
import kotlinx.android.synthetic.main.item_favorite_article_view.view.*
import kotlinx.android.synthetic.main.item_load_more_view.view.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.greenrobot.eventbus.EventBus
import top.easelink.framework.base.BaseViewHolder
import top.easelink.framework.threadpool.CommonPool
import top.easelink.framework.threadpool.Main
import top.easelink.lcg.R
import top.easelink.lcg.ui.main.model.OpenArticleEvent
import top.easelink.lcg.ui.main.source.local.ArticlesLocalDataSource.delArticleFromFavorite
import top.easelink.lcg.ui.main.source.model.ArticleEntity

class FavoriteArticlesAdapter(private var favoriteArticlesViewModel: FavoriteArticlesViewModel) :
    RecyclerView.Adapter<BaseViewHolder>() {
    private val mArticleEntities: MutableList<ArticleEntity> = mutableListOf()

    override fun getItemCount(): Int {
        return if (mArticleEntities.isEmpty()) {
            1
        } else {
            mArticleEntities.size + 1
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (mArticleEntities.isEmpty()) {
            VIEW_TYPE_EMPTY
        } else {
            if (position == mArticleEntities.size) {
                VIEW_TYPE_LOAD_MORE
            } else {
                VIEW_TYPE_NORMAL
            }
        }
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.onBind(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return when (viewType) {
            VIEW_TYPE_NORMAL -> ArticleViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    R.layout.item_favorite_article_view, parent, false)
            )
            VIEW_TYPE_LOAD_MORE -> LoadMoreViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    R.layout.item_load_more_view, parent, false
                )
            )
            VIEW_TYPE_EMPTY -> EmptyViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    R.layout.item_favorite_article_empty_view, parent, false
                )
            )
            else -> EmptyViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    R.layout.item_favorite_article_empty_view, parent, false
                )
            )
        }
    }

    fun addItems(articleEntityList: List<ArticleEntity>) {
        mArticleEntities.addAll(articleEntityList)
        notifyDataSetChanged()
    }

    fun clearItems() {
        mArticleEntities.clear()
    }

    inner class ArticleViewHolder internal constructor(private val view: View) : BaseViewHolder(view) {

        // will be assigned in onBind(), should not be called before onBind()
        private lateinit var articleEntity: ArticleEntity

        private fun onItemClick() {
            val event = OpenArticleEvent(articleEntity.url)
            EventBus.getDefault().post(event)
        }

        override fun onBind(position: Int) {
            articleEntity = mArticleEntities[position].also {entity ->
                view.remove_button.setOnClickListener {
                    GlobalScope.launch(CommonPool) {
                        //TODO del url has a confirmation button to click, support this in the future
//                        if (entity.delUrl.isNotEmpty()) {
//                            val doc = Client.sendGetRequestWithQuery(entity.delUrl)
//                        }
                        if (delArticleFromFavorite(entity.id)) {
                            mArticleEntities.remove(entity)
                            GlobalScope.launch(Main) {
                                notifyItemRemoved(position)
                                notifyItemRangeChanged(position, mArticleEntities.size - position)
                            }
                        }
                    }
                }
            }
            view.run {
                favorite_container.apply {
                    when {
                        // odd
                        position % 2 == 0 -> {
                            setBackgroundResource(R.drawable.rectangle_bg_slightly_light_gray_4dp)
                        }
                        // even
                        else -> {
                            setBackgroundResource(R.drawable.rectangle_bg_white_4dp)
                        }
                    }
                }.setOnClickListener {
                    onItemClick()
                }
                title_text_view.text = articleEntity.title
                if (articleEntity.author.isNotBlank()) {
                    author_text_view.apply {
                        text = articleEntity.author
                        visibility = View.VISIBLE
                    }
                }  else {
                    author_text_view.visibility = View.GONE
                }
            }
        }

    }

    inner class EmptyViewHolder internal constructor(val view: View) :
        BaseViewHolder(view) {
        override fun onBind(position: Int) {
            view.sync_favorites.setOnClickListener {
                favoriteArticlesViewModel.syncFavorites()
            }
        }
    }

    inner class LoadMoreViewHolder internal constructor(val view: View) : BaseViewHolder(view) {
        override fun onBind(position: Int) {
            view.loading.visibility = View.VISIBLE
            favoriteArticlesViewModel.fetchArticles(ArticleFetcher.FetchType.FETCH_MORE) {
                view.post {
                    view.loading.visibility = View.GONE
                }
            }
        }
    }

    companion object {
        private const val VIEW_TYPE_EMPTY = 0
        private const val VIEW_TYPE_NORMAL = 1
        private const val VIEW_TYPE_LOAD_MORE = 2
    }
}